File: apps/queue/service/models/queue_definition.go
// Copyright 2023-2026 Ant Investor Ltd
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package models

import "github.com/pitabwire/frame/data"

// QueueDefinition represents a named queue with configuration.
type QueueDefinition struct {
	data.BaseModel `gorm:"embedded"`

	Name           string `gorm:"column:name;not null"`
	Description    string `gorm:"column:description"`
	Active         bool   `gorm:"column:active;not null;default:true"`
	PriorityLevels int    `gorm:"column:priority_levels;not null;default:3"`
	MaxCapacity    int    `gorm:"column:max_capacity;not null;default:0"`
	SLAMinutes     int    `gorm:"column:sla_minutes;not null;default:30"`
	Config         string `gorm:"column:config;type:jsonb"`
}

// TableName returns the database table name.
func (QueueDefinition) TableName() string {
	return "queue_definitions"
}

File: ui_queuestore/lib/src/models/queue_definition.dart
/// A queue definition with its configuration.
class QueueDefinition {
  const QueueDefinition({
    required this.id,
    required this.name,
    this.description = '',
    this.active = true,
    this.priorityLevels = 3,
    this.maxCapacity = 0,
    this.slaMinutes = 30,
    this.config,
    this.createdAt,
    this.modifiedAt,
  });

  final String id;
  final String name;
  final String description;
  final bool active;
  final int priorityLevels;
  final int maxCapacity;
  final int slaMinutes;
  final Map<String, dynamic>? config;
  final DateTime? createdAt;
  final DateTime? modifiedAt;

  factory QueueDefinition.fromJson(Map<String, dynamic> json) {
    return QueueDefinition(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      description: json['description'] as String? ?? '',
      active: json['active'] as bool? ?? true,
      priorityLevels: json['priority_levels'] as int? ?? 3,
      maxCapacity: json['max_capacity'] as int? ?? 0,
      slaMinutes: json['sla_minutes'] as int? ?? 30,
      config: json['config'] as Map<String, dynamic>?,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
      modifiedAt: json['modified_at'] != null
          ? DateTime.tryParse(json['modified_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'priority_levels': priorityLevels,
      'max_capacity': maxCapacity,
      'sla_minutes': slaMinutes,
      if (config != null) 'config': config,
    };
  }

  QueueDefinition copyWith({
    String? name,
    String? description,
    bool? active,
    int? priorityLevels,
    int? maxCapacity,
    int? slaMinutes,
    Map<String, dynamic>? config,
  }) {
    return QueueDefinition(
      id: id,
      name: name ?? this.name,
      description: description ?? this.description,
      active: active ?? this.active,
      priorityLevels: priorityLevels ?? this.priorityLevels,
      maxCapacity: maxCapacity ?? this.maxCapacity,
      slaMinutes: slaMinutes ?? this.slaMinutes,
      config: config ?? this.config,
      createdAt: createdAt,
      modifiedAt: modifiedAt,
    );
  }
}
